File: December2022/Java/BinaryTreeMaximumPathSum.java
// @saorav21994

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    /* Max of [ root , root.left + root, root.right + root, root + root.left + root.right ]  */


    int res = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        findMax(root);

        return res;
    }


    public int findMax(TreeNode root) {
        
        if (root == null)
            return 0;

        int left = findMax(root.left);
        int right = findMax(root.right);    

        int max1 = Math.max(root.val, root.val + Math.max(left, right));
        int max2 = Math.max(max1, root.val+left+right);

        res = Math.max(res, max2);

        return max1;

    }

}

File: December2022/C++/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
private:
    
    int findMaxPathSum(TreeNode* root, int &maxSum){
        
        if(root == NULL){
            
            return 0;
        }
        
       
        int node = root->val;
        
        int leftMaxSum  = max(0, findMaxPathSum(root->left, maxSum));
        int rightMaxSum = max(0, findMaxPathSum(root->right, maxSum));
        
        maxSum = max( maxSum, leftMaxSum + rightMaxSum + node);
  
        return max(leftMaxSum + node, rightMaxSum + node);
    }
    
public:
    int maxPathSum(TreeNode* root) {
        
        int maxSum = INT_MIN;
        
        findMaxPathSum(root, maxSum);
        
        return maxSum;
    }
};